// proxies/index.mjs
import {SQSClient} from '@aws-sdk/client-sqs';
import {SendMessageCommand} from '@aws-sdk/client-sqs';
import {DynamoDBClient, GetItemCommand} from '@aws-sdk/client-dynamodb';

export const sqsClient = new SQSClient();
export const dynamoClient = new DynamoDBClient();

const capitalizeTwoWordsRequestQueue = `${process.env.SQS_BASE_URL}/capitalizeTwoWords-request-queue`;

export async function capitalizeTwoWords(_ctx)  {
  const argsString = JSON.stringify(
    Array.from(arguments).slice(0, capitalizeTwoWords.length)
  );

  return request(capitalizeTwoWordsRequestQueue, argsString);
}

const capitalizeTwoWords_1RequestQueue = `${process.env.SQS_BASE_URL}/capitalizeTwoWords_1-request-queue`;

export async function capitalizeTwoWords_1(_ctx)  {
  const argsString = JSON.stringify(
    Array.from(arguments).slice(0, capitalizeTwoWords_1.length)
  );

  return request(capitalizeTwoWords_1RequestQueue, argsString);
}

const capitalizeTwoWords_2RequestQueue = `${process.env.SQS_BASE_URL}/capitalizeTwoWords_2-request-queue`;

export async function capitalizeTwoWords_2(_ctx)  {
  const argsString = JSON.stringify(
    Array.from(arguments).slice(0, capitalizeTwoWords_2.length)
  );

  return request(capitalizeTwoWords_2RequestQueue, argsString);
}

const capitalizeWordRequestQueue = `${process.env.SQS_BASE_URL}/capitalizeWord-request-queue`;

export async function capitalizeWord(_ctx)  {
  const argsString = JSON.stringify(
    Array.from(arguments).slice(0, capitalizeWord.length)
  );

  return request(capitalizeWordRequestQueue, argsString);
}

async function request(requestQueueUrl, argsString) {
  const requestId = crypto.randomUUID();

  console.log(`Sending request with ID ${requestId}`);

  await sqsClient.send(new SendMessageCommand({
    QueueUrl: requestQueueUrl,
    MessageAttributes: {
      OriginRequestId: {
        DataType: 'String',
        StringValue: requestId
      }
    },
    MessageBody: argsString
  }));

  for (let attempt = 0; attempt < 30; attempt++) {
    console.log(`Waiting for response, attempt ${attempt + 1}...`);

    try {
      const response = await dynamoClient.send(new GetItemCommand({
        TableName: 'response-table',
        Key: {
          'request-id': { S: requestId }
        }
      }));

      if (response.Item) {
        console.log("Found response in DynamoDB:", response.Item);
        const result = response.Item.result.S;
        return JSON.parse(result);
      }
    } catch (error) {
      console.log("Error checking DynamoDB:", error);
    }

    // Wait before next attempt
    await new Promise(resolve => setTimeout(resolve, 2000));
  }

  throw new Error("Did not receive a response for requestId: " + requestId);
}
