// proxies/index.mjs
import {SQSClient} from '@aws-sdk/client-sqs';
import {DeleteMessageCommand, SendMessageCommand, ReceiveMessageCommand} from '@aws-sdk/client-sqs';

export const sqsClient = new SQSClient();

const capitalizeWordsRequestQueue = `${process.env.SQS_BASE_URL}/capitalizeWords-request-queue`;
const capitalizeWordsResponseQueue = `${process.env.SQS_BASE_URL}/capitalizeWords-response-queue`;

export async function capitalizeWords(string)  {
  const argsString = JSON.stringify(
    Array.from(arguments).slice(0, capitalizeWords.length)
  );

  return request(
    capitalizeWordsRequestQueue, 
    capitalizeWordsResponseQueue, 
    argsString
  );
}

const capitalizeWordRequestQueue = `${process.env.SQS_BASE_URL}/capitalizeWord-request-queue`;
const capitalizeWordResponseQueue = `${process.env.SQS_BASE_URL}/capitalizeWord-response-queue`;

export async function capitalizeWord(word)  {
  const argsString = JSON.stringify(
    Array.from(arguments).slice(0, capitalizeWord.length)
  );

  return request(
    capitalizeWordRequestQueue, 
    capitalizeWordResponseQueue, 
    argsString
  );
}

async function request(requestQueueUrl, responseQueueUrl, argsString) {
  const requestId = crypto.randomUUID();

  console.log(`Sending request with ID ${requestId}`);

  await sqsClient.send(new SendMessageCommand({
    QueueUrl: requestQueueUrl,
    MessageAttributes: {
      OriginRequestId: {
        DataType: 'String',
        StringValue: requestId
      }
    },
    MessageBody: argsString
  }));

  for (let attempt = 0; attempt < 5; attempt++) {
    console.log(`Waiting for response, attempt ${attempt + 1}...`);
    const receiveResponse = await sqsClient.send(new ReceiveMessageCommand({
      QueueUrl: responseQueueUrl,
      MessageAttributeNames: ['OriginRequestId'],
      WaitTimeSeconds: 3,
      MaxNumberOfMessages: 10,
      VisibilityTimeout: 1
    }));

    if (receiveResponse.Messages) {
      console.log("Received messages: ", receiveResponse.Messages);
      for (const message of receiveResponse.Messages) {
        console.log("Message: ", message);

        if (message.MessageAttributes.OriginRequestId.StringValue === requestId) {
          const result = message.Body;

          await sqsClient.send(new DeleteMessageCommand({
            QueueUrl: responseQueueUrl,
            ReceiptHandle: message.ReceiptHandle
          }));

          console.log("Received relevant message, deleting: ", message.ReceiptHandle);

          return JSON.parse(result);
        }
      }
    }
  }

  throw new Error("Did not receive a response for requestId: " + requestId);
}
